dropbear: bump to 2025.88

rockdrilla opened the pull request:
- update dropbear to latest stable 2025.88;
  for the changes see https://matt.ucc.asn.au/dropbear/CHANGES
- rewrite `100-pubkey_path.patch`
- refresh remaining patches

- configuration is heavily rewritten: fine-grained configuration of almost all things in Dropbear
- disable RSA-SHA1 by default
- new post-quantum algorithms are available: `sntrup761` and `mlkem768`
- `sntrup761` is enabled by default for devices with sufficient flash space (`!SMALL_FLASH`)
- `mlkem768` is NOT enabled by default (adds ~64k to `dropbear` binary itself and ~20k to `.apk` package)

Compilation results:

- default target: mips, ath79/generic

  - Dropbear 2024.86:
    binary size: 264292
    package size: 118606

  - Dropbear 2025.88:
    binary size: 264352 (+60 bytes)
    package size: 124433 (+5827 bytes)

- "small" target: mips, ath79/tiny (`FEATURES += low_mem small_flash`)

  - Dropbear 2024.86:
    binary size: 197077
    package size: 95498

  - Dropbear 2025.88:
    binary size: 197061 (-64 bytes)
    package size: 94217 (-1281 bytes)

NB: end users should migrate their `CONFIG_DROPBEAR_MODERN_ONLY=y` manually:

```
# CONFIG_DROPBEAR_AES128 is not set
# CONFIG_DROPBEAR_AES256 is not set
# CONFIG_DROPBEAR_LEGACY_COMPAT is not set
# CONFIG_DROPBEAR_RSA is not set
```

and optionally set `CONFIG_DROPBEAR_MLKEM768=y` and/or `CONFIG_DROPBEAR_SNTRUP761=y` for post-quantum resistance.

Neustradamus commented:
@rockdrilla: Good job! :)

rockdrilla commented:
Compile-/run-tested: mediatek/filogic (GL.iNet GL-MT6000)

Size changes:

- binary package (apk):  112849 -> 135709 (+22860)
- binary package (opkg): 112638 -> 135408 (+22770)
- `/usr/sbin/dropbear`:  263099 -> 263179 (+80)

stklcode commented:
Successfully build and run tested on ath79/generic (Ubiquiti AirCube AC)

For run-test backported to 24.10.1

Size changes:
* binary package (apk): 117376 -> 144327 (+26951)
* binary package (opkg): 116838 -> 144108 (+27270)
* `/usr/sbin/dropbear`: 264308 -> 330116 (+65808)

Sure your +80 is correct despite the +22k on the package? :thinking:

rockdrilla commented:
@stklcode I've already removed staging directory with build results and can't verify it again right now.

schuettecarsten commented:
This update breaks login on my device:

```
Wed Apr 30 22:20:51 2025 authpriv.info dropbear[6460]: Child connection from 192.168.10.4:41942
Wed Apr 30 22:20:51 2025 authpriv.info dropbear[6460]: /etc must be owned by user or root, and not writable by group or others
Wed Apr 30 22:20:51 2025 authpriv.info dropbear[6460]: Exit before auth from <192.168.10.4:41942>: (user 'root', 0 fails): Exited normally
```

rockdrilla commented:
@schuettecarsten are you able to provide somehow command output?

```sh
ls -ld /etc /etc/dropbear /etc/dropbear/*
```

Here is mine:

```sh
root@OpenWrt:~# ls -ld /etc /etc/dropbear /etc/dropbear/*
drwxr-xr-x 1 root root 3488 Apr 29 19:38 /etc
drwxr-xr-x 1 root root 3488 Jan  1  1970 /etc/dropbear
-rw------- 1 root root   81 Nov 19  2022 /etc/dropbear/authorized_keys
-rw------- 1 root root  141 Apr 26 18:00 /etc/dropbear/dropbear_ecdsa_host_key
-rw------- 1 root root   83 Nov 19  2022 /etc/dropbear/dropbear_ed25519_host_key
root@OpenWrt:~# dropbear -V
Dropbear v2025.87
```

Personal config:
```sh
$ grep -Fi dropbear krdwrt-flint2.cfg 
# CONFIG_DROPBEAR_AGENTFORWARD is not set
CONFIG_DROPBEAR_ASKPASS=y
# CONFIG_DROPBEAR_DBCLIENT_AGENTFORWARD is not set
CONFIG_DROPBEAR_ECC=y
CONFIG_DROPBEAR_ECDSA_SK=y
CONFIG_DROPBEAR_MODERN_ONLY=y
```

schuettecarsten commented:
I already fixed that using `chmod 755 /etc`.
But I wonder why my `/etc` was different and worked before.

```
root@GatewayDummi:/etc# ls -ld /etc /etc/dropbear /etc/dropbear/*
drwxr-xr-x    1 root     root          3488 Apr 30 07:45 /etc
drwxr-xr-x    1 root     root          3488 Apr 29 22:47 /etc/dropbear
-rw-------    1 root     root           188 Apr 22 07:53 /etc/dropbear/authorized_keys
-rw-------    1 root     root           140 Dec  4  2023 /etc/dropbear/dropbear_ecdsa_host_key
-rw-------    1 root     root            83 Oct 19  2023 /etc/dropbear/dropbear_ed25519_host_key
-rw-------    1 root     root           805 Oct 19  2023 /etc/dropbear/dropbear_rsa_host_key
```

My config, grepped for `dropbear`:

```
CONFIG_DROPBEAR_CURVE25519=y
CONFIG_DROPBEAR_ECC=y
CONFIG_DROPBEAR_ECC_FULL=y
CONFIG_DROPBEAR_ED25519=y
CONFIG_DROPBEAR_CHACHA20POLY1305=y
CONFIG_DROPBEAR_U2F=y
CONFIG_DROPBEAR_ECDSA_SK=y
CONFIG_DROPBEAR_ED25519_SK=y
CONFIG_DROPBEAR_ZLIB=y
CONFIG_DROPBEAR_DBCLIENT=y
# CONFIG_DROPBEAR_ASKPASS is not set
CONFIG_DROPBEAR_DBCLIENT_AGENTFORWARD=y
CONFIG_DROPBEAR_SCP=y
CONFIG_DROPBEAR_AGENTFORWARD=y
# CONFIG_DROPBEAR_MODERN_ONLY is not set
```

rockdrilla commented:
@schuettecarsten it was working before because there was (my) previous version of `100-pubkey_path.patch`. :roll_eyes: 
Permission handling in upstream was rewritten a lot and I'd to rewrite this patch too with no changes on further checks.

I could add some lines to init script to fix permissions of `/etc` directory itself but this is somewhat undesirable in that package.

More logical way is to open issue against `base-files` package: `boot` service (`/etc/init.d/boot`) is very early script to start, so permission fixups are natural to it.

rockdrilla commented:
I've changed a bit `100-pubkey_path.patch`:

```diff
- if (strncmp(svr_opts.authorized_keys_dir, "/", 1) == 0) break;
+ if (svr_opts.authorized_keys_dir[0] == '/') break;
```

IMO this is a bit more optimal way.

Neustradamus commented:
@rockdrilla: Dropbear 2025.88 has been released.

rockdrilla commented:
@Neustradamus ok, thanks, will refresh asap.

rockdrilla commented:
I'm currently working on adapting #15460 into #18606, stay tuned.

rockdrilla commented:
I've added `050-dropbear-multihop-fix.patch` in order to fix build with `proxycmd` and without `netcat` (reported to upstream: mkj/dropbear#363).

rockdrilla commented:
NB: these numbers are harvested without any further changes in package.

Package size (with .apk):

```
2024.86: 118636
2025.88: 146152
diff: +27516
```

Binary size itself:

```
2024.86: 264292
2025.88: 330104
diff: +65812
```

Reason: Dropbear 2025.87 introduced two post-quantum KEX - `mlkem768` and `sntrup761` (jfyi, they're internally reuse some code from `curve25519`/`ed25519`). They're unintentionally enabled by default (for good reasons).

Hovewer, we'd better to control features to be built or not. I'll adapt #15460 (with many many thanks to @M95D) and adopt options from Dropbear 2025.87 and 2025.88.

kwinz commented:
I am really looking forward to `sntrup761x25519-sha512` and/or `mlkem768x25519-sha256` (ML-KEM) being enabled by default on OpenWRT with `dropbear 2025.88` being merged, so I can finally remove the non post-quantum safe key exchange exceptions for my routers from my clients' `ssh_config`.

rockdrilla commented:
I've done preliminary adoption of #15460.
I'd like to ask you to correct spelling, wording, _et cetera_.

Proposals and bug reports are welcome!

rockdrilla commented:
Preliminary results:

default target: mips, ath79/generic

Dropbear 2024.86:
binary size: 264292
package size: 118630

Dropbear 2025.88:
binary size: 264204 (-88 bytes)
package size: 116726 (-1904 bytes)

"small" target: mips, ath79/tiny (`FEATURES += low_mem small_flash`)

Dropbear 2024.86:
binary size: 197077
package size: 95551

Dropbear 2025.88:
binary size: 197013 (-64 bytes)
package size: 92217 (-3334 bytes)

rockdrilla commented:
I've accidentally disabled by default Curve25519 (sorry).

rockdrilla commented:
Friendly ping @Neustradamus, @stklcode, @schuettecarsten, @kwinz

@hauke, @rsalvaterra, @Borromini PTAL

stklcode commented:
Good job :+1: 

Successfully build and run tested on ath79/generic (Ubiquiti AirCube AC) on 24.10.1 with default flags, post-quantum enabled and "modern".

Binary size with default flags -12, with mlkem768 enabled +65k, with mlkem768 and "modern" +80.

danpawlik reviewed (approved):
Works fine on BPi-R4 and GL.B3000

rockdrilla commented:
Rebased on top of current HEAD.

rockdrilla commented:
@robimarko JFYI: I need this MR to be merged first in order to work further on glibc 2.41.

kwinz commented:
Could you please try to enable at least one of the post quantum KEX by default?

Since binary size is a concern I think it would be sufficient to enable only `DROPBEAR_SNTRUP761`, since that one is more widely supported by clients (since OpenSSH 8.9).

rockdrilla commented:
I couldn't estimate proposed changes right now (and corresponding size changes) but I thought about it. More likely I would suggest enabling both PQ algorithms for `!SMALL_FLASH` devices.

robimarko commented on package/network/services/dropbear/patches/050-dropbear-multihop-fix.patch:
This needs a proper header and description of what it does

rockdrilla commented:
Changes:
- enable `mlkem768` and `sntrup761` for `!SMALL_FLASH`
- write notes for `patches/050-dropbear-multihop-fix.patch`

To be done:
- (fresh) binary/package size estimation

aparcar commented:
Hi, thanks for your worked! I compared the activated ciphers and it seems that `ssh-rsa` is disabled, even when legacy is enabled?

Dropbear of SNAPSHOTS

```json
{
  "kex_algorithms": [
    "curve25519-sha256",
    "curve25519-sha256@libssh.org",
    "diffie-hellman-group14-sha256",
    "kexguess2@matt.ucc.asn.au",
    "kex-strict-s-v00@openssh.com"
  ],
  "server_host_key_algorithms": ["ssh-ed25519", "rsa-sha2-256", "ssh-rsa"],
  "encryption_algorithms": [
    "chacha20-poly1305@openssh.com",
    "aes128-ctr",
    "aes256-ctr"
  ],
  "mac_algorithms": ["hmac-sha2-256"]
}
```

Dropbear after the update

```shell
{
  "kex_algorithms": [
    "sntrup761x25519-sha512",
    "sntrup761x25519-sha512@openssh.com",
    "mlkem768x25519-sha256",
    "curve25519-sha256",
    "curve25519-sha256@libssh.org",
    "diffie-hellman-group14-sha256",
    "kexguess2@matt.ucc.asn.au",
    "kex-strict-s-v00@openssh.com"
  ],
  "server_host_key_algorithms": ["ssh-ed25519", "rsa-sha2-256"],
  "encryption_algorithms": [
    "chacha20-poly1305@openssh.com",
    "aes256-ctr",
    "aes128-ctr"
  ],
  "mac_algorithms": ["hmac-sha2-256"]
}
```

rockdrilla commented:
@aparcar afaik "ssh-rsa" is for "RSA with SHA-1" whereas SHA-1 is disabled by default in both upstream and OpenWrt. The one may build it's own dropbear package with necessary legacy options being enabled.

stklcode commented:
IMHO that’s fine, the legacy one should be disabled by default while the RSA-SHA256 can be still around for a now.

SHA1 must be enable explicitly and dependes on LEGACY_COMPAT, but not vice versa.

To be discussed if this should be configured backward-compatible in case of backport to a stable branch, but in _main_ I’d vote for the non-legacy approach.

```
	config DROPBEAR_RSA_SHA1
		bool "RSA-SHA1 [WEAK]"
		depends on DROPBEAR_LEGACY_COMPAT
		help
			This enables the following public key algorithm:
			 * ssh-rsa
			 ...

	config DROPBEAR_RSA
		bool "RSA-SHA256"
		default y

		help
			This enables the following public key algorithm:
			 * rsa-sha2-256
			 ...
```

hauke commented on package/network/services/dropbear/Makefile:
```diff
@@ -109,24 +114,33 @@ CONFIGURE_ARGS += \
 	COMPAT_USER_SHELLS,"/bin/ash","/bin/sh" \
 	DEFAULT_PATH,"$(TARGET_INIT_PATH)" \
 	DEFAULT_ROOT_PATH,"$(TARGET_INIT_PATH)" \
-	DROPBEAR_DSS,0 \
-	DROPBEAR_CLI_NETCAT,0 \
 	DO_MOTD,0 \
-	DROPBEAR_DH_GROUP14_SHA1,0 \
-	DROPBEAR_SHA1_HMAC,0 \
```
Are these 4 removed configuration options removed from upstream dropbear?

hauke commented on package/network/services/dropbear/files/dropbear.init:
```diff
@@ -177,7 +177,7 @@ validate_section_dropbear()
 		'IdleTimeout:uinteger:0' \
 		'MaxAuthTries:uinteger:3' \
-		'RecvWindowSize:uinteger:262144' \
+		'RecvWindowSize:uinteger:0' \
```
Why do you change this back? It was set in 865ae1c10c65001813413da95eb1b8cd06a7e1c1.

rockdrilla replied:
We've discussed that change here: https://github.com/openwrt/openwrt/pull/15460#issuecomment-2108069634
I agree with @M95D.

rockdrilla replied:
One more moment: default value is compile-time constant defined with `CONFIG_DROPBEAR_DEFAULT_RECV_WINDOW` which is `32768` for low-memory devices and `262144` otherwise. The one may change this value either via build config (compile-time) or via uci config (run-time). IMO it seems legit.

rockdrilla commented on package/network/services/dropbear/Makefile:
```diff
@@ -87,16 +93,15 @@ endef
 
 CONFIGURE_ARGS += \
 	--disable-pam \
-	--enable-openpty \
```
`openpty` and `syslog` are enabled by default, there's no need to specify them explicitly.

rockdrilla replied:
From build log:
```
configure: Using openpty if available
checking for library containing openpty... none required
configure: Enabling syslog
```

aparcar commented:
@rockdrilla you you kindly create a single commit that explicitly disables SHA1 so anyone going through the commit history sees this clearly? Apart from that I'm happy to merge this once the discussion with @hauke is cleared.

rockdrilla committed:
dropbear: bump to 2025.88

- update dropbear to latest stable 2025.88;
  for the changes see https://matt.ucc.asn.au/dropbear/CHANGES
- rewrite 100-pubkey_path.patch
- refresh remaining patches

Signed-off-by: Konstantin Demin <rockdrilla@gmail.com>

rockdrilla committed:
dropbear: refactor configuration

This allows to fine-tune dropbear build options.

This change is heavily based on similar work done by Marius Dinu earlier
so I'd like to say many thanks to original author.

Signed-off-by: Konstantin Demin <rockdrilla@gmail.com>

rockdrilla committed:
dropbear: disable RSA-SHA1 by default

Upstream has disabled SHA-1 algorithms by default since version 2025.87.
SHA-1 has known weakness and most SSH implementations support alternatives.

Signed-off-by: Konstantin Demin <rockdrilla@gmail.com>

rockdrilla committed:
dropbear: simplify rebuild on config changes

It's hard or even impossible to track affected sources
so it's safe to remove all built objects (if any).

Signed-off-by: Konstantin Demin <rockdrilla@gmail.com>

Neustradamus commented:
@aparcar: Thanks for merging commits!

@rockdrilla, good job, thanks!

Merged commits:
- https://github.com/openwrt/openwrt/commit/bbe4d6ddb2a97271a779521304fb1e26c70246fe
- https://github.com/openwrt/openwrt/commit/d8ff2d8cba7c90d7d77c493cf585142758346009
- https://github.com/openwrt/openwrt/commit/3dc05213e0261bb960006c957c32f8f14559abe9
- https://github.com/openwrt/openwrt/commit/5e0fbca9b98e9bc415bbaa9aeeecb95848699a3f

aparcar commented on package/network/services/dropbear/Config.in:
Excellent Cath, do you mind creating a PR?
